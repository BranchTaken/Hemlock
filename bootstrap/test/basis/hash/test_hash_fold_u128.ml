open! Basis.Rudiments
open! Basis
open Hash

let u128_of_tup = RudimentsInt0.u128_of_tup

let test () =
  let hash_fold u128s t = begin
    State.Gen.init t
    |> State.Gen.fold_u128 Stdlib.(Int64.of_int (Array.length u128s)) ~f:(fun i ->
      Stdlib.(Array.get u128s Int64.(to_int i))
    )
    |> State.Gen.fini
  end in
  let rec test_hash_fold u128s_list = begin
    match u128s_list with
    | [] -> ()
    | u128s :: u128s_list' -> begin
        File.Fmt.stdout
        |> Fmt.fmt "hash_fold "
        |> (Array.pp pp) u128s
        |> Fmt.fmt " -> "
        |> pp (t_of_state State.(hash_fold u128s empty))
        |> Fmt.fmt "\n"
        |> ignore;
        test_hash_fold u128s_list'
      end
  end in
  (* These test inputs were manually verified against the reference MurmurHash3 implementation. *)
  let u128s_list = [
    [||];

    [|u128_of_tup (0x0123456789abcdefL, 0xfedcba9876543210L)|];

    [|U128.zero|];

    [|u128_of_tup (0xfedcba9876543210L, 0x0123456789abcdefL)|];

    [|u128_of_tup (0x0123456789abcdefL, 0xfedcba9876543210L);
      U128.zero;
      u128_of_tup (0xfedcba9876543210L, 0x0123456789abcdefL)|]
  ] in
  test_hash_fold u128s_list

let _ = test ()
