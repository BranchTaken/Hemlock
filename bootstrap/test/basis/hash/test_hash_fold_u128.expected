hash_fold [||] -> 0x0000_0000_0000_0000_0000_0000_0000_0000u128
hash_fold [|0xfedc_ba98_7654_3210_0123_4567_89ab_cdefu128|] -> 0x6bc0_bad4_001a_c79a_3a7b_e286_a34a_7a71u128
hash_fold [|0x0000_0000_0000_0000_0000_0000_0000_0000u128|] -> 0xb465_a9ec_cd79_1cb6_4bbd_1bf2_7da9_18d6u128
hash_fold [|0x0123_4567_89ab_cdef_fedc_ba98_7654_3210u128|] -> 0x0eb8_4470_9565_7807_4f87_4a06_b0db_435eu128
hash_fold [|0xfedc_ba98_7654_3210_0123_4567_89ab_cdefu128; 0x0000_0000_0000_0000_0000_0000_0000_0000u128; 0x0123_4567_89ab_cdef_fedc_ba98_7654_3210u128|] -> 0x368d_c0ef_bb0a_0838_afb5_c175_06c0_c15eu128
